exchanges/stream/websocket.go

```diff
@@ -784,16 +1016,38 @@ func (w *Websocket) GetName() string {
 
 // GetChannelDifference finds the difference between the subscribed channels
 // and the new subscription list when pairs are disabled or enabled.
-func (w *Websocket) GetChannelDifference(newSubs subscription.List) (sub, unsub subscription.List) {
-	if w.subscriptions == nil {
-		w.subscriptions = subscription.NewStore()
+func (w *Websocket) GetChannelDifference(conn Connection, newSubs subscription.List) (sub, unsub subscription.List) {
+	var subscriptionStore **subscription.Store
+	if candidate, ok := w.connections[conn]; ok {
+		subscriptionStore = &candidate.Subscriptions
+	} else {
+		subscriptionStore = &w.subscriptions
+	}
+	if *subscriptionStore == nil {
+		*subscriptionStore = subscription.NewStore()
 	}
-	return w.subscriptions.Diff(newSubs)
+	return (*subscriptionStore).Diff(newSubs)
 }
 
 // UnsubscribeChannels unsubscribes from a list of websocket channel
-func (w *Websocket) UnsubscribeChannels(channels subscription.List) error {
-	if w.subscriptions == nil || len(channels) == 0 {
+func (w *Websocket) UnsubscribeChannels(conn Connection, channels subscription.List) error {
+	if len(channels) == 0 {
+		return nil // No channels to unsubscribe from is not an error
+	}
+
+	if candidate, ok := w.connections[conn]; ok {
+		if candidate.Subscriptions == nil {
+			return nil // No channels to unsubscribe from is not an error
+		}
+		for _, s := range channels {
+			if candidate.Subscriptions.Get(s) == nil {
+				return fmt.Errorf("%w: %s", subscription.ErrNotFound, s)
+			}
+		}
+		return candidate.Setup.Unsubscriber(context.TODO(), conn, channels)
```

Comment:
It's hard to consolidate due to the unsubscriber method parameters from the original to the new multi connection signature. The originals will be removed eventually when everything is using the multi connection manager. Maybe I am missing what you are asking, here's what I did just to make:

```go
// UnsubscribeChannels unsubscribes from a list of websocket channel
func (w *Websocket) UnsubscribeChannels(conn Connection, channels subscription.List) error {
	if len(channels) == 0 {
		return nil // No channels to unsubscribe from is not an error
	}
	if candidate, ok := w.connections[conn]; ok {
		return w.unsubscribe(candidate.Subscriptions, channels, func(channels subscription.List) error {
			return candidate.Setup.Unsubscriber(context.TODO(), conn, channels)
		})
	}
	return w.unsubscribe(w.subscriptions, channels, func(channels subscription.List) error {
		return w.Unsubscriber(channels)
	})
}

func (w *Websocket) unsubscribe(store *subscription.Store, channels subscription.List, unsubscribe func(channels subscription.List) error) error {
	if store == nil {
		return nil // No channels to unsubscribe from is not an error
	}
	for _, s := range channels {
		if err := store.Remove(s); err != nil {
			if store.Get(s) == nil {
				return fmt.Errorf("%w: %s", subscription.ErrNotFound, s)
			}
		}
	}
	return unsubscribe(channels)
}
```

So the foot print is pretty much unchanged

Reply:
It's less about footprint and more about repeated code. I definitely didn't  _make_ something, I just spotted two sections doing the same thing and thought it could be abstracted. What you posted looks nice. I can give a more thorough look on my next pass